File: 88-policy-object/PolicyDemo.java
/*
 * Ngôn ngữ: Java
 * Công dụng: Policy object — luật nghiệp vụ đổi theo NGỮ CẢNH (quốc gia, hạng khách)
 * mà code gọi không đổi một chữ. Bài cho nổ ba con bug: chuỗi if-else chép ở ba nơi rồi
 * một nơi quên mất nước Đức nên thu thuế 0%; thiếu chính sách thì âm thầm rơi về 0 thay
 * vì báo lỗi; và bùng nổ tổ hợp 4×3 = 12 lớp khi hai trục luật bị trộn làm một.
 * Tại sao cần học: bài 87 ghép luật LÚC VIẾT CODE. Bài này chọn luật LÚC CHẠY — đó là
 * toàn bộ khác biệt, và nó là thứ quyết định một hệ thống có mở rộng sang thị trường
 * mới trong một ngày hay trong một quý. Riêng Java cho một công cụ mạnh ở đây: `enum`
 * liệt kê được TẤT CẢ ngữ cảnh, nên "đã đủ chính sách cho mọi quốc gia chưa" là một
 * bài test chạy được, không phải một câu hỏi lúc review.
 */
import java.util.EnumMap;
import java.util.LinkedHashMap;
import java.util.Map;

public class PolicyDemo {

    enum QuocGia { VN, JP, US, DE }
    enum HangKhach { THUONG, BAC, VANG }

    // =====================================================================
    // POLICY — một luật, được CHỌN lúc chạy
    // =====================================================================
    interface ChinhSachThue {
        long tinhThue(long tienHang);
        String moTa();                       // có tên, đọc lên thành câu (bài 81)
    }

    record ThueTheoTiLe(int phanTram, String ten) implements ChinhSachThue {
        @Override public long tinhThue(long tienHang) { return tienHang * phanTram / 100; }
        @Override public String moTa() { return ten + " " + phanTram + "%"; }
    }
    record MienThue(String ten) implements ChinhSachThue {
        @Override public long tinhThue(long tienHang) { return 0; }
        @Override public String moTa() { return ten; }
    }

    // Bảng tra: THÊM MỘT THỊ TRƯỜNG = THÊM MỘT DÒNG, không sửa dòng nào.
    static final Map<QuocGia, ChinhSachThue> BANG_THUE = new EnumMap<>(QuocGia.class);
    static {
        BANG_THUE.put(QuocGia.VN, new ThueTheoTiLe(10, "VAT Việt Nam"));
        BANG_THUE.put(QuocGia.JP, new ThueTheoTiLe(8, "thuế tiêu dùng Nhật"));
        BANG_THUE.put(QuocGia.US, new MienThue("không thuế liên bang"));
        BANG_THUE.put(QuocGia.DE, new ThueTheoTiLe(19, "USt Đức"));
    }

    // Tra chính sách: THIẾU thì NỔ, không âm thầm về 0. Xem phần 3.
    static ChinhSachThue chinhSachCho(QuocGia q) {
        ChinhSachThue cs = BANG_THUE.get(q);
        if (cs == null) throw new IllegalStateException("chưa có chính sách thuế cho " + q);
        return cs;
    }

    // =====================================================================
    // TRỤC THỨ HAI — giảm giá theo hạng khách, ĐỘC LẬP với thuế
    // =====================================================================
    interface ChinhSachGiamGia { long tinhGiam(long tienHang); }

    static final Map<HangKhach, ChinhSachGiamGia> BANG_GIAM = new EnumMap<>(HangKhach.class);
    static {
        BANG_GIAM.put(HangKhach.THUONG, t -> 0);
        BANG_GIAM.put(HangKhach.BAC, t -> t * 5 / 100);
        BANG_GIAM.put(HangKhach.VANG, t -> t * 10 / 100);
    }

    /** Tầng ứng dụng chỉ GHÉP hai trục lại. Nó không biết nước nào bao nhiêu phần trăm. */
    static long tinhTongPhaiTra(long tienHang, QuocGia quocGia, HangKhach hang) {
        long giam = BANG_GIAM.get(hang).tinhGiam(tienHang);
        long sauGiam = tienHang - giam;
        return sauGiam + chinhSachCho(quocGia).tinhThue(sauGiam);
    }

    // ---- Self-check ----
    private static void check(boolean ok, String msg) {
        if (!ok) throw new AssertionError(msg);
    }

    public static void main(String[] args) {
        // ---- 1. CON BUG: chuỗi if-else chép ở ba nơi ----
        // Ba nơi cùng cần thuế: màn hình thanh toán, sinh hoá đơn, báo cáo doanh thu.
        // Nước Đức được thêm vào tháng trước, và chỉ hai trong ba nơi được cập nhật.
        java.util.function.BiFunction<QuocGia, Long, Long> thanhToan = (q, t) ->
                q == QuocGia.VN ? t * 10 / 100 : q == QuocGia.JP ? t * 8 / 100
                        : q == QuocGia.DE ? t * 19 / 100 : 0L;
        java.util.function.BiFunction<QuocGia, Long, Long> hoaDon = (q, t) ->
                q == QuocGia.VN ? t * 10 / 100 : q == QuocGia.JP ? t * 8 / 100
                        : q == QuocGia.DE ? t * 19 / 100 : 0L;
        java.util.function.BiFunction<QuocGia, Long, Long> baoCao = (q, t) ->
                q == QuocGia.VN ? t * 10 / 100 : q == QuocGia.JP ? t * 8 / 100 : 0L;
        //                                                        ^ QUÊN nước Đức

        long tien = 100_000_000L;
        check(thanhToan.apply(QuocGia.DE, tien) == 19_000_000L, "thanh toán thu đúng 19%");
        check(hoaDon.apply(QuocGia.DE, tien) == 19_000_000L, "hoá đơn ghi đúng 19%");
        check(baoCao.apply(QuocGia.DE, tien) == 0L, "báo cáo ghi 0% — lệch 19 triệu mỗi đơn");
        check(thanhToan.apply(QuocGia.DE, tien) - baoCao.apply(QuocGia.DE, tien) == 19_000_000L,
                "sổ sách và tiền thật không khớp nhau");
        // Không ngoại lệ, không cảnh báo. Nhánh `else` trả 0 nuốt trọn lỗi — và với thuế
        // thì 0 là một con số hoàn toàn hợp lệ (nước Mỹ đúng là 0%), nên không ai nghi ngờ.

        // ---- 2. POLICY: một nguồn sự thật, ba nơi cùng dùng ----
        check(chinhSachCho(QuocGia.DE).tinhThue(tien) == 19_000_000L, "một bảng, một câu trả lời");
        check(chinhSachCho(QuocGia.US).tinhThue(tien) == 0L, "Mỹ 0% — nhưng là 0% CÓ TÊN");
        check(chinhSachCho(QuocGia.US).moTa().equals("không thuế liên bang"),
                "và tên đó phân biệt được với 'chưa cấu hình'");
        // Đây là điểm tinh tế nhất của bài: `MienThue` và "thiếu cấu hình" đều cho ra 0,
        // nhưng một cái là QUYẾT ĐỊNH NGHIỆP VỤ còn cái kia là LỖI. Chuỗi if-else không
        // phân biệt được hai thứ đó; bảng chính sách thì có.

        // ---- 3. ĐIỀU CHỈ `enum` LÀM ĐƯỢC: KIỂM TRA ĐỦ CHÍNH SÁCH BẰNG MÁY ----
        for (QuocGia q : QuocGia.values()) {
            check(BANG_THUE.containsKey(q), "thiếu chính sách thuế cho " + q);
        }
        check(BANG_THUE.size() == QuocGia.values().length, "đủ 4/4 quốc gia");
        // Bốn dòng trên là một bài test chạy trong CI. Thêm `QuocGia.FR` vào enum mà quên
        // thêm dòng vào bảng -> test đỏ NGAY, trước khi có đơn hàng nào từ Pháp.
        //
        // Với chuỗi if-else thì không có cách nào viết bài test tương đương, vì không có
        // gì để mà liệt kê — nhánh `else` luôn "xử lý được" mọi giá trị.

        // ---- 4. THIẾU CHÍNH SÁCH PHẢI NỔ, KHÔNG ĐƯỢC ÂM THẦM VỀ 0 ----
        Map<QuocGia, ChinhSachThue> bangThieu = new LinkedHashMap<>(BANG_THUE);
        bangThieu.remove(QuocGia.DE);
        long thueImLang = bangThieu.getOrDefault(QuocGia.DE, new MienThue("mặc định")).tinhThue(tien);
        check(thueImLang == 0L, "getOrDefault: thiếu chính sách -> 0 đồng, không ai biết");

        boolean noLen = false;
        try { chinhSachCho(null); } catch (IllegalStateException e) { noLen = true; }
        check(noLen, "tra chính sách phải NỔ khi thiếu — 19 triệu không được im lặng biến mất");
        // `getOrDefault(..., mặc định)` là một trong những dòng nguy hiểm nhất trong mã
        // nghiệp vụ. Null Object (bài 64) chỉ đúng khi "không có gì" là hành vi HỢP LỆ.
        // Với thuế thì không: thiếu chính sách là tin xấu, và tin xấu phải kêu to.

        // ---- 5. HAI TRỤC ĐỘC LẬP: 4 + 3, KHÔNG PHẢI 4 × 3 ----
        // Cám dỗ: một lớp cho mỗi tổ hợp — `ThueVnKhachVang`, `ThueDeKhachBac`, ...
        int soLopNeuTronTruc = QuocGia.values().length * HangKhach.values().length;
        int soLopKhiTachTruc = QuocGia.values().length + HangKhach.values().length;
        check(soLopNeuTronTruc == 12 && soLopKhiTachTruc == 7, "12 lớp so với 7");
        // Và con số đó nổ theo cấp số nhân: thêm trục thứ ba (kênh bán: online/đại lý/
        // cửa hàng) thì 12 -> 36, còn 7 -> 10. Quy tắc: mỗi TRỤC BIẾN THIÊN là một bảng
        // chính sách riêng, và tầng ứng dụng ghép chúng lại (bài 63 · decorator là một
        // cách ghép khác cho cùng bài toán này).

        check(tinhTongPhaiTra(100_000L, QuocGia.VN, HangKhach.THUONG) == 110_000L,
                "VN thường: 100.000 + 10% = 110.000");
        check(tinhTongPhaiTra(100_000L, QuocGia.VN, HangKhach.VANG) == 99_000L,
                "VN vàng: giảm 10% còn 90.000, +10% thuế = 99.000");
        check(tinhTongPhaiTra(100_000L, QuocGia.US, HangKhach.VANG) == 90_000L,
                "Mỹ vàng: giảm 10%, không thuế");
        // Chú ý THỨ TỰ: giảm giá TRƯỚC, thuế SAU — thuế tính trên số tiền thực trả. Đây
        // là một luật nghiệp vụ, và nó nằm ở tầng ứng dụng vì nó nói về QUAN HỆ giữa hai
        // chính sách chứ không thuộc chính sách nào. Đảo thứ tự là sai luật thuế ở hầu
        // hết các nước — và đó là loại bug không ai phát hiện cho tới lúc bị kiểm toán.

        // ---- 6. THÊM THỊ TRƯỜNG MỚI: ĐO SỐ CHỖ PHẢI SỬA ----
        // Với if-else: sửa 3 nhánh (thanh toán, hoá đơn, báo cáo) — và bài học ở phần 1
        // là khả năng quên một chỗ không phải giả thuyết.
        // Với policy: thêm MỘT dòng vào `BANG_THUE`, và bài test ở phần 3 canh giúp.
        Map<QuocGia, ChinhSachThue> bangMoRong = new LinkedHashMap<>(BANG_THUE);
        int truoc = bangMoRong.size();
        bangMoRong.put(QuocGia.US, new ThueTheoTiLe(7, "thuế bang California"));  // đổi luật Mỹ
        check(bangMoRong.size() == truoc, "sửa luật MỘT nước: đúng một dòng, không đụng nước khác");
        check(bangMoRong.get(QuocGia.VN).tinhThue(tien) == 10_000_000L, "Việt Nam không hề hấn gì");

        // ---- 7. POLICY vs STRATEGY vs SPECIFICATION ----
        //
        //   Mẫu           | Trả lời câu hỏi          | Chọn lúc nào  | Ví dụ ở đây
        //   --------------|--------------------------|---------------|------------------
        //   Specification | "có thoả mãn không?"     | ghép lúc viết | duocVayTinChap (87)
        //   Policy        | "luật ở ngữ cảnh này?"   | tra LÚC CHẠY  | BANG_THUE
        //   Strategy      | "làm bằng cách nào?"     | tra lúc chạy  | thuật toán nén/sắp xếp
        //
        // Policy và Strategy có HÌNH DẠNG giống hệt nhau — cùng là một interface với
        // nhiều cài đặt và một chỗ chọn. Khác nhau ở Ý ĐỊNH: strategy đổi CÁCH LÀM cho
        // cùng một kết quả (sắp xếp nhanh hay chậm, kết quả như nhau); policy đổi CHÍNH
        // KẾT QUẢ vì nghiệp vụ ở ngữ cảnh đó khác. Nhầm lẫn hai thứ không gây bug, nhưng
        // gọi đúng tên giúp người sau biết được phép đổi cái gì mà không phá gì.
        check(chinhSachCho(QuocGia.VN).tinhThue(1000) != chinhSachCho(QuocGia.JP).tinhThue(1000),
                "policy: hai ngữ cảnh, hai KẾT QUẢ khác nhau — và cả hai đều đúng");

        // ---- 8. RANH GIỚI: khoá tra chính sách phải là KIỂU CỦA MIỀN ----
        // `Map<String, ChinhSachThue>` với khoá `"VN"`, `"vn"`, `"VNM"` là cách chắc chắn
        // nhất để có một bug không ai tìm ra. Dùng `enum` (hoặc value object) thì:
        //   - gõ sai là lỗi biên dịch;
        //   - liệt kê được hết -> viết được bài test ở phần 3;
        //   - IDE tìm được mọi nơi dùng.
        // Chuỗi chỉ nên xuất hiện ở BIÊN (đọc file cấu hình, nhận request), và được đổi
        // sang enum ngay tại đó (bài 76 · fail fast, bài 78 · DTO mapping).
        check(QuocGia.valueOf("DE") == QuocGia.DE, "biên đổi chuỗi -> enum một lần, ngay lúc vào");

        System.out.println("OK");
    }
}

File: 88-policy-object/policy_demo.py
# Ngôn ngữ: Python
# Công dụng: Bản Python của cùng bài học — luật đổi theo ngữ cảnh (quốc gia, hạng
# khách) mà code gọi không đổi. Ba con bug: chuỗi if-else chép ba nơi rồi một nơi quên
# nước Đức (lệch 19 triệu mỗi đơn); thiếu chính sách thì âm thầm về 0; và bùng nổ tổ
# hợp 4×3 khi trộn hai trục luật.
# Tại sao cần học: Python cho cách đăng ký chính sách gọn nhất trong ba ngôn ngữ — một
# decorator `@dang_ky(QuocGia.DE)` và bảng tra tự điền. Nhưng chính sự gọn đó mở ra cái
# bẫy riêng của Python: bảng chỉ được điền khi MODULE chứa chính sách đó được import.
# Một chính sách nằm trong file chưa ai import là một chính sách không tồn tại — và hệ
# thống sẽ chạy êm với thuế 0%. Bài đo đúng chỗ đó, và dựng bài test chặn nó.

from enum import Enum, auto
from typing import Callable, Dict


class QuocGia(Enum):
    VN = auto()
    JP = auto()
    US = auto()
    DE = auto()


class HangKhach(Enum):
    THUONG = auto()
    BAC = auto()
    VANG = auto()


# =====================================================================
# POLICY — một luật, được CHỌN lúc chạy
# =====================================================================
class ChinhSachThue:
    def tinh_thue(self, tien_hang: int) -> int:
        raise NotImplementedError

    def mo_ta(self) -> str:              # có tên, đọc lên thành câu (bài 81)
        raise NotImplementedError


class ThueTheoTiLe(ChinhSachThue):
    def __init__(self, phan_tram, ten):
        self.phan_tram, self.ten = phan_tram, ten

    def tinh_thue(self, tien_hang):
        return tien_hang * self.phan_tram // 100

    def mo_ta(self):
        return f"{self.ten} {self.phan_tram}%"


class MienThue(ChinhSachThue):
    def __init__(self, ten):
        self.ten = ten

    def tinh_thue(self, tien_hang):
        return 0

    def mo_ta(self):
        return self.ten


BANG_THUE: Dict[QuocGia, ChinhSachThue] = {}


def dang_ky(quoc_gia):
    """Decorator: THÊM MỘT THỊ TRƯỜNG = THÊM MỘT LỚP, không sửa dòng nào."""
    def gan(lop):
        BANG_THUE[quoc_gia] = lop()
        return lop
    return gan


@dang_ky(QuocGia.VN)
class ThueVN(ThueTheoTiLe):
    def __init__(self):
        super().__init__(10, "VAT Việt Nam")


@dang_ky(QuocGia.JP)
class ThueJP(ThueTheoTiLe):
    def __init__(self):
        super().__init__(8, "thuế tiêu dùng Nhật")


@dang_ky(QuocGia.US)
class ThueUS(MienThue):
    def __init__(self):
        super().__init__("không thuế liên bang")


@dang_ky(QuocGia.DE)
class ThueDE(ThueTheoTiLe):
    def __init__(self):
        super().__init__(19, "USt Đức")


def chinh_sach_cho(quoc_gia, bang=None):
    """Tra chính sách: THIẾU thì NỔ, không âm thầm về 0. Xem phần 4."""
    bang = BANG_THUE if bang is None else bang
    cs = bang.get(quoc_gia)
    if cs is None:
        raise RuntimeError(f"chưa có chính sách thuế cho {quoc_gia}")
    return cs


# =====================================================================
# TRỤC THỨ HAI — giảm giá theo hạng khách, ĐỘC LẬP với thuế
# =====================================================================
BANG_GIAM: Dict[HangKhach, Callable[[int], int]] = {
    HangKhach.THUONG: lambda t: 0,
    HangKhach.BAC: lambda t: t * 5 // 100,
    HangKhach.VANG: lambda t: t * 10 // 100,
}


def tinh_tong_phai_tra(tien_hang, quoc_gia, hang):
    """Tầng ứng dụng chỉ GHÉP hai trục lại. Nó không biết nước nào bao nhiêu phần trăm."""
    giam = BANG_GIAM[hang](tien_hang)
    sau_giam = tien_hang - giam
    return sau_giam + chinh_sach_cho(quoc_gia).tinh_thue(sau_giam)


# ---- Self-check ----
if __name__ == "__main__":
    tien = 100_000_000

    # ---- 1. CON BUG: chuỗi if-else chép ở ba nơi ----
    # Ba nơi cùng cần thuế: màn hình thanh toán, sinh hoá đơn, báo cáo doanh thu. Nước
    # Đức được thêm tháng trước, và chỉ hai trong ba nơi được cập nhật.
    def thanh_toan(q, t):
        if q is QuocGia.VN:
            return t * 10 // 100
        if q is QuocGia.JP:
            return t * 8 // 100
        if q is QuocGia.DE:
            return t * 19 // 100
        return 0

    hoa_don = thanh_toan

    def bao_cao(q, t):
        if q is QuocGia.VN:
            return t * 10 // 100
        if q is QuocGia.JP:
            return t * 8 // 100
        return 0                                # <- QUÊN nước Đức

    assert thanh_toan(QuocGia.DE, tien) == 19_000_000, "thanh toán thu đúng 19%"
    assert hoa_don(QuocGia.DE, tien) == 19_000_000, "hoá đơn ghi đúng 19%"
    assert bao_cao(QuocGia.DE, tien) == 0, "báo cáo ghi 0% — lệch 19 triệu mỗi đơn"
    assert thanh_toan(QuocGia.DE, tien) - bao_cao(QuocGia.DE, tien) == 19_000_000, \
        "sổ sách và tiền thật không khớp nhau"
    # Nhánh `return 0` nuốt trọn lỗi — và với thuế thì 0 là một con số hoàn toàn hợp lệ
    # (nước Mỹ đúng là 0%), nên không ai nghi ngờ.

    # ---- 2. POLICY: một nguồn sự thật, ba nơi cùng dùng ----
    assert chinh_sach_cho(QuocGia.DE).tinh_thue(tien) == 19_000_000, "một bảng, một câu trả lời"
    assert chinh_sach_cho(QuocGia.US).tinh_thue(tien) == 0, "Mỹ 0% — nhưng là 0% CÓ TÊN"
    assert chinh_sach_cho(QuocGia.US).mo_ta() == "không thuế liên bang", \
        "và tên đó phân biệt được với 'chưa cấu hình'"
    # Điểm tinh tế nhất của bài: `MienThue` và "thiếu cấu hình" đều cho ra 0, nhưng một
    # cái là QUYẾT ĐỊNH NGHIỆP VỤ còn cái kia là LỖI. Chuỗi if-else không phân biệt được.

    # ---- 3. KIỂM TRA ĐỦ CHÍNH SÁCH BẰNG MÁY ----
    thieu = [q for q in QuocGia if q not in BANG_THUE]
    assert thieu == [], f"thiếu chính sách thuế cho: {thieu}"
    assert len(BANG_THUE) == len(QuocGia), "đủ 4/4 quốc gia"
    # Ba dòng trên là một bài test chạy trong CI. Thêm `QuocGia.FR` vào enum mà quên viết
    # lớp chính sách -> test đỏ NGAY, trước khi có đơn hàng nào từ Pháp.
    #
    # Với chuỗi if-else thì không có cách nào viết bài test tương đương, vì không có gì
    # để mà liệt kê — nhánh `return 0` luôn "xử lý được" mọi giá trị.

    # ---- 4. CÁI BẪY RIÊNG CỦA PYTHON: bảng chỉ đầy khi module ĐƯỢC IMPORT ----
    # `@dang_ky(...)` chỉ chạy lúc định nghĩa lớp, tức là lúc module chứa nó được import.
    # Đặt `ThueDE` trong `chinh_sach/duc.py` mà không ai import file đó -> `BANG_THUE`
    # thiếu `DE`, và hệ thống chạy êm với... thuế 0%.
    bang_mo_phong_thieu_import = {q: cs for q, cs in BANG_THUE.items() if q is not QuocGia.DE}
    assert QuocGia.DE not in bang_mo_phong_thieu_import, "mô phỏng: file chính sách Đức chưa ai import"
    no_len = False
    try:
        chinh_sach_cho(QuocGia.DE, bang_mo_phong_thieu_import)
    except RuntimeError:
        no_len = True
    assert no_len, "tra chính sách phải NỔ khi thiếu — 19 triệu không được im lặng biến mất"

    thue_im_lang = bang_mo_phong_thieu_import.get(QuocGia.DE, MienThue("mặc định")).tinh_thue(tien)
    assert thue_im_lang == 0, "`.get(..., mặc định)` là dòng nguy hiểm nhất trong mã nghiệp vụ"
    # Null Object (bài 64) chỉ đúng khi "không có gì" là hành vi HỢP LỆ. Với thuế thì
    # không: thiếu chính sách là tin xấu, và tin xấu phải kêu to.
    #
    # Cách chặn cái bẫy import: bài test ở phần 3 phải chạy SAU khi import gói chính sách
    # (`import chinh_sach` trong `__init__.py`), và tốt nhất là gói đó tự nạp mọi module
    # con bằng `pkgutil.iter_modules`. Nếu không thì chính bài test cũng sẽ không thấy
    # thiếu gì — nó chỉ kiểm tra được những gì đã được import.

    # ---- 5. HAI TRỤC ĐỘC LẬP: 4 + 3, KHÔNG PHẢI 4 × 3 ----
    so_lop_neu_tron_truc = len(QuocGia) * len(HangKhach)
    so_lop_khi_tach_truc = len(QuocGia) + len(HangKhach)
    assert (so_lop_neu_tron_truc, so_lop_khi_tach_truc) == (12, 7), "12 lớp so với 7"
    # Và con số đó nổ theo cấp số nhân: thêm trục thứ ba (kênh bán) thì 12 -> 36, còn
    # 7 -> 10. Quy tắc: mỗi TRỤC BIẾN THIÊN là một bảng chính sách riêng, và tầng ứng
    # dụng ghép chúng lại (bài 63 · decorator là một cách ghép khác cho cùng bài toán).

    assert tinh_tong_phai_tra(100_000, QuocGia.VN, HangKhach.THUONG) == 110_000, \
        "VN thường: 100.000 + 10% = 110.000"
    assert tinh_tong_phai_tra(100_000, QuocGia.VN, HangKhach.VANG) == 99_000, \
        "VN vàng: giảm 10% còn 90.000, +10% thuế = 99.000"
    assert tinh_tong_phai_tra(100_000, QuocGia.US, HangKhach.VANG) == 90_000, \
        "Mỹ vàng: giảm 10%, không thuế"
    # Chú ý THỨ TỰ: giảm giá TRƯỚC, thuế SAU — thuế tính trên số tiền thực trả. Đây là
    # một luật nghiệp vụ, và nó nằm ở tầng ứng dụng vì nó nói về QUAN HỆ giữa hai chính
    # sách chứ không thuộc chính sách nào. Đảo thứ tự là sai luật thuế ở hầu hết các
    # nước — loại bug không ai phát hiện cho tới lúc bị kiểm toán.

    # ---- 6. THÊM/SỬA THỊ TRƯỜNG: ĐO SỐ CHỖ PHẢI SỬA ----
    # Với if-else: sửa 3 nhánh — và bài học ở phần 1 là khả năng quên một chỗ không phải
    # giả thuyết. Với policy: thêm MỘT lớp, và bài test ở phần 3 canh giúp.
    bang_mo_rong = dict(BANG_THUE)
    truoc = len(bang_mo_rong)
    bang_mo_rong[QuocGia.US] = ThueTheoTiLe(7, "thuế bang California")   # đổi luật Mỹ
    assert len(bang_mo_rong) == truoc, "sửa luật MỘT nước: đúng một dòng, không đụng nước khác"
    assert bang_mo_rong[QuocGia.VN].tinh_thue(tien) == 10_000_000, "Việt Nam không hề hấn gì"

    # ---- 7. POLICY vs STRATEGY vs SPECIFICATION ----
    #
    #   Mẫu           | Trả lời câu hỏi          | Chọn lúc nào  | Ví dụ ở đây
    #   --------------|--------------------------|---------------|------------------
    #   Specification | "có thoả mãn không?"     | ghép lúc viết | duoc_vay_tin_chap (87)
    #   Policy        | "luật ở ngữ cảnh này?"   | tra LÚC CHẠY  | BANG_THUE
    #   Strategy      | "làm bằng cách nào?"     | tra lúc chạy  | thuật toán nén/sắp xếp
    #
    # Policy và Strategy có HÌNH DẠNG giống hệt nhau. Khác nhau ở Ý ĐỊNH: strategy đổi
    # CÁCH LÀM cho cùng một kết quả; policy đổi CHÍNH KẾT QUẢ vì nghiệp vụ ở ngữ cảnh đó
    # khác. Nhầm lẫn không gây bug, nhưng gọi đúng tên giúp người sau biết được phép đổi
    # cái gì mà không phá gì.
    assert chinh_sach_cho(QuocGia.VN).tinh_thue(1000) != chinh_sach_cho(QuocGia.JP).tinh_thue(1000), \
        "policy: hai ngữ cảnh, hai KẾT QUẢ khác nhau — và cả hai đều đúng"

    # ---- 8. RANH GIỚI: khoá tra chính sách phải là KIỂU CỦA MIỀN ----
    # `{"VN": ..., "vn": ..., "VNM": ...}` là cách chắc chắn nhất để có một bug không ai
    # tìm ra. Dùng `Enum` thì gõ sai là `KeyError` ngay, liệt kê được hết (phần 3), và
    # IDE tìm được mọi nơi dùng. Chuỗi chỉ nên xuất hiện ở BIÊN (đọc cấu hình, nhận
    # request) và được đổi sang enum ngay tại đó (bài 76 · fail fast, bài 78 · DTO).
    assert QuocGia["DE"] is QuocGia.DE, "biên đổi chuỗi -> enum một lần, ngay lúc vào"
    sai_khoa = False
    try:
        QuocGia["Duc"]
    except KeyError:
        sai_khoa = True
    assert sai_khoa, "gõ sai mã nước là nổ ngay tại biên, không trôi vào trong"

    print("OK")
